File: atosim/org.palladiosimulator.blockchainsystems.threesim_plugin/src/main/java/org/palladiosimulator/blockchainsystems/threesim_plugin/ThreesimTab.java
package org.palladiosimulator.blockchainsystems.threesim_plugin;

import org.eclipse.debug.core.ILaunchConfiguration;
import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
import org.eclipse.debug.ui.AbstractLaunchConfigurationTab;
import org.eclipse.jface.layout.GridLayoutFactory;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.widgets.Combo;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Group;
import org.eclipse.swt.widgets.Label;
import org.palladiosimulator.blockchainsystems.plugin.ui.abstractions.TextField;
import org.palladiosimulator.blockchainsystems.plugin.utils.DoubleVerifier;
import org.palladiosimulator.blockchainsystems.plugin.utils.LongVerifier;

public class ThreesimTab extends AbstractLaunchConfigurationTab {

    private static final double MIN_FAILURE_THROUGHPUT_THRESHOLD = 0.0;
    private static final double MIN_SHANNON_ENTROPY_K = 0.0;
    private static final double MIN_NAKAMOTO_COEFFICIENT_THRESHOLD = 0.0;
    private static final double MAX_NAKAMOTO_COEFFICIENT_THRESHOLD = 100.0;
    private static final double MIN_RELIABILITY_OBSERVATION_TIMESPAN = 0.0;

    private boolean isInitialized = false;

    private TextField failureThroughputThresholdField;
    private TextField shannonEntropyKField;
    private TextField nakamotoCoefficientThresholdField;
    private TextField reliabilityObservationTimespanField;
    private TextField blockIntervalField;
    private TextField propagationDelayField;
    private TextField nodeDegreeField;
    private TextField maxBlockSizeField;
    private TextField networkBandwidthField;
    private TextField transactionADelayField;
    private TextField transactionBAccelerationField;
    private Combo combinedAttackModeCombo;

    @Override
    public void createControl(Composite parent) {
        Composite root = new Composite(parent, SWT.BORDER);
        GridLayoutFactory.swtDefaults().numColumns(1).applyTo(root);

        Group group = new Group(root, SWT.NONE);
        group.setText("3SIM Parameters");
        GridLayoutFactory.swtDefaults().numColumns(3).spacing(0, 10).applyTo(group);
        group.setLayoutData(new GridData(SWT.FILL, SWT.BEGINNING, true, false));

        failureThroughputThresholdField = new TextField(group,
                "Failure Throughput Threshold:", " transactions/min",
                DoubleVerifier.INSTANCE,
                ThreesimAttributes.FAILURE_THROUGHPUT_THRESHOLD,
                ThreesimAttributes.FAILURE_THROUGHPUT_THRESHOLD_DEFAULT,
                s -> parseDouble(s, v -> v >= MIN_FAILURE_THROUGHPUT_THRESHOLD));

        shannonEntropyKField = new TextField(group,
                "Shannon Entropy K:", "",
                DoubleVerifier.INSTANCE,
                ThreesimAttributes.SHANNON_ENTROPY_K,
                ThreesimAttributes.SHANNON_ENTROPY_K_DEFAULT,
                s -> parseDouble(s, v -> v > MIN_SHANNON_ENTROPY_K));

        nakamotoCoefficientThresholdField = new TextField(group,
                "Nakamoto Coefficient Threshold:", " %",
                DoubleVerifier.INSTANCE,
                ThreesimAttributes.NAKAMOTO_COEFFICIENT_THRESHOLD,
                ThreesimAttributes.NAKAMOTO_COEFFICIENT_THRESHOLD_DEFAULT,
                s -> parseDouble(s, v -> v >= MIN_NAKAMOTO_COEFFICIENT_THRESHOLD
                        && v <= MAX_NAKAMOTO_COEFFICIENT_THRESHOLD));

        reliabilityObservationTimespanField = new TextField(group,
                "Reliability Observation Timespan:", " hours",
                DoubleVerifier.INSTANCE,
                ThreesimAttributes.RELIABILITY_OBSERVATION_TIMESPAN,
                ThreesimAttributes.RELIABILITY_OBSERVATION_TIMESPAN_DEFAULT,
                s -> parseDouble(s, v -> v > MIN_RELIABILITY_OBSERVATION_TIMESPAN));

        blockIntervalField = new TextField(group,
                "Block Interval:", " ms",
                DoubleVerifier.INSTANCE,
                ThreesimAttributes.BLOCK_INTERVAL,
                ThreesimAttributes.BLOCK_INTERVAL_DEFAULT,
                s -> parseDouble(s, v -> v > 0));

        propagationDelayField = new TextField(group,
                "Propagation Delay:", " ms",
                DoubleVerifier.INSTANCE,
                ThreesimAttributes.PROPAGATION_DELAY,
                ThreesimAttributes.PROPAGATION_DELAY_DEFAULT,
                s -> parseDouble(s, v -> v >= 0));

        nodeDegreeField = new TextField(group,
                "Node Degree:", "",
                LongVerifier.INSTANCE,
                ThreesimAttributes.NODE_DEGREE,
                ThreesimAttributes.NODE_DEGREE_DEFAULT,
                s -> parseLong(s, v -> v > 0));

        maxBlockSizeField = new TextField(group,
                "Max Block Size:", " bytes",
                LongVerifier.INSTANCE,
                ThreesimAttributes.MAX_BLOCK_SIZE,
                ThreesimAttributes.MAX_BLOCK_SIZE_DEFAULT,
                s -> parseLong(s, v -> v > 0));

        networkBandwidthField = new TextField(group,
                "Network Bandwidth:", " bytes/s",
                DoubleVerifier.INSTANCE,
                ThreesimAttributes.NETWORK_BANDWIDTH,
                ThreesimAttributes.NETWORK_BANDWIDTH_DEFAULT,
                s -> parseDouble(s, v -> v > 0));

        transactionADelayField = new TextField(group,
                "Transaction A Delay:", " ms",
                LongVerifier.INSTANCE,
                ThreesimAttributes.TRANSACTION_A_DELAY,
                ThreesimAttributes.TRANSACTION_A_DELAY_DEFAULT,
                s -> parseLong(s, v -> v >= 0));

        transactionBAccelerationField = new TextField(group,
                "Transaction B Acceleration:", " ms",
                LongVerifier.INSTANCE,
                ThreesimAttributes.TRANSACTION_B_ACCELERATION,
                ThreesimAttributes.TRANSACTION_B_ACCELERATION_DEFAULT,
                s -> parseLong(s, v -> v >= 0));

        Label combinedAttackLabel = new Label(group, SWT.NONE);
        combinedAttackLabel.setText("Combined Attack Mode:");

        combinedAttackModeCombo = new Combo(group, SWT.DROP_DOWN | SWT.READ_ONLY);
        combinedAttackModeCombo.setItems("None", "Selfish + Race", "Selfish + Finney", "Selfish + Lead Stubborn", "Selfish + Trail Stubborn");
        combinedAttackModeCombo.select(0);
        combinedAttackModeCombo.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false));

        new Label(group, SWT.NONE).setText("");

        combinedAttackModeCombo.addModifyListener(e -> updateLaunchConfigurationDialog());

        setControl(root);
        isInitialized = true;
    }

    @Override
    public String getName() {
        return "ATO-SIM Parameters";
    }

    @Override
    public void initializeFrom(ILaunchConfiguration configuration) {
        if (!isInitialized) return;
        try {
            failureThroughputThresholdField.initializeFrom(configuration);
            shannonEntropyKField.initializeFrom(configuration);
            nakamotoCoefficientThresholdField.initializeFrom(configuration);
            reliabilityObservationTimespanField.initializeFrom(configuration);
            blockIntervalField.initializeFrom(configuration);
            propagationDelayField.initializeFrom(configuration);
            nodeDegreeField.initializeFrom(configuration);
            maxBlockSizeField.initializeFrom(configuration);
            networkBandwidthField.initializeFrom(configuration);
            transactionADelayField.initializeFrom(configuration);
            transactionBAccelerationField.initializeFrom(configuration);

            String mode = configuration.getAttribute(
                    ThreesimAttributes.COMBINED_ATTACK_MODE,
                    ThreesimAttributes.COMBINED_ATTACK_MODE_DEFAULT);
            if (ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_RACE.equals(mode)) {
                combinedAttackModeCombo.select(1);
            } else if (ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_FINNEY.equals(mode)) {
                combinedAttackModeCombo.select(2);
            } else if (ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_LEAD_STUBBORN.equals(mode)) {
                combinedAttackModeCombo.select(3);
            } else if (ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_TRAIL_STUBBORN.equals(mode)) {
                combinedAttackModeCombo.select(4);
            } else {
                combinedAttackModeCombo.select(0);
            }
        } catch (Exception ignored) {
        }
    }

    @Override
    public void performApply(ILaunchConfigurationWorkingCopy configuration) {
        if (!isInitialized) return;
        failureThroughputThresholdField.performApply(configuration);
        shannonEntropyKField.performApply(configuration);
        nakamotoCoefficientThresholdField.performApply(configuration);
        reliabilityObservationTimespanField.performApply(configuration);
        blockIntervalField.performApply(configuration);
        propagationDelayField.performApply(configuration);
        nodeDegreeField.performApply(configuration);
        maxBlockSizeField.performApply(configuration);
        networkBandwidthField.performApply(configuration);
        transactionADelayField.performApply(configuration);
        transactionBAccelerationField.performApply(configuration);

        String combinedMode = switch (combinedAttackModeCombo.getSelectionIndex()) {
            case 1 -> ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_RACE;
            case 2 -> ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_FINNEY;
            case 3 -> ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_LEAD_STUBBORN;
            case 4 -> ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_TRAIL_STUBBORN;
            default -> ThreesimAttributes.COMBINED_ATTACK_MODE_NONE;
        };
        configuration.setAttribute(ThreesimAttributes.COMBINED_ATTACK_MODE, combinedMode);
    }

    @Override
    public void setDefaults(ILaunchConfigurationWorkingCopy configuration) {
        if (!isInitialized) return;
        failureThroughputThresholdField.setDefaults(configuration);
        shannonEntropyKField.setDefaults(configuration);
        nakamotoCoefficientThresholdField.setDefaults(configuration);
        reliabilityObservationTimespanField.setDefaults(configuration);
        blockIntervalField.setDefaults(configuration);
        propagationDelayField.setDefaults(configuration);
        nodeDegreeField.setDefaults(configuration);
        maxBlockSizeField.setDefaults(configuration);
        networkBandwidthField.setDefaults(configuration);
        transactionADelayField.setDefaults(configuration);
        transactionBAccelerationField.setDefaults(configuration);
        configuration.setAttribute(ThreesimAttributes.COMBINED_ATTACK_MODE,
                ThreesimAttributes.COMBINED_ATTACK_MODE_DEFAULT);
    }

    @Override
    public void activated(ILaunchConfigurationWorkingCopy workingCopy) {
        super.activated(workingCopy);
        updateLaunchConfigurationDialog();
    }

    @Override
    public boolean isValid(ILaunchConfiguration launchConfig) {
        if (!isInitialized) return false;
        return failureThroughputThresholdField.isValid()
                && shannonEntropyKField.isValid()
                && nakamotoCoefficientThresholdField.isValid()
                && reliabilityObservationTimespanField.isValid()
                && blockIntervalField.isValid()
                && propagationDelayField.isValid()
                && nodeDegreeField.isValid()
                && maxBlockSizeField.isValid()
                && networkBandwidthField.isValid()
                && transactionADelayField.isValid()
                && transactionBAccelerationField.isValid();
    }

    private interface DoubleCheck {
        boolean test(double v);
    }

    private interface LongCheck {
        boolean test(long v);
    }

    private static boolean parseDouble(String s, DoubleCheck check) {
        try {
            return check.test(Double.parseDouble(s));
        } catch (NumberFormatException e) {
            return false;
        }
    }

    private static boolean parseLong(String s, LongCheck check) {
        try {
            return check.test(Long.parseLong(s));
        } catch (NumberFormatException e) {
            return false;
        }
    }
}

File: atosim/org.palladiosimulator.blockchainsystems.threesim_plugin/src/main/kotlin/org/palladiosimulator/blockchainsystems/threesim_plugin/ThreesimTab.kt
package org.palladiosimulator.blockchainsystems.threesim_plugin

import org.eclipse.debug.core.ILaunchConfiguration
import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy
import org.eclipse.debug.ui.AbstractLaunchConfigurationTab
import org.eclipse.jface.layout.GridLayoutFactory
import org.eclipse.swt.SWT
import org.eclipse.swt.layout.GridData
import org.eclipse.swt.widgets.Composite
import org.eclipse.swt.widgets.Group
import org.eclipse.swt.widgets.Combo
import org.eclipse.swt.widgets.Label
import org.palladiosimulator.blockchainsystems.plugin.ui.abstractions.TextField
import org.palladiosimulator.blockchainsystems.plugin.utils.DoubleVerifier
import org.palladiosimulator.blockchainsystems.plugin.utils.LongVerifier
import org.palladiosimulator.blockchainsystems.threesim_plugin.ThreesimAttributes

/**
 * Tab for configuring Threesim-specific simulation parameters.
 *
 * @author Davis Riedel
 */
class ThreesimTab : AbstractLaunchConfigurationTab() {
  companion object {
    private const val MIN_RELIABILITY_OBSERVATION_TIMESPAN = 0.0
    private const val MIN_NAKAMOTO_COEFFICIENT_THRESHOLD = 0.0
    private const val MAX_NAKAMOTO_COEFFICIENT_THRESHOLD = 100.0
    private const val MIN_SHANNON_ENTROPY_K = 0.0
    private const val MIN_FAILURE_THROUGHPUT_THRESHOLD = 0.0
  }

  private var isInitialized = false

  private lateinit var failureThroughputThresholdField: TextField
  private lateinit var shannonEntropyKField: TextField
  private lateinit var nakamotoCoefficientThresholdField: TextField
  private lateinit var reliabilityObservationTimespanField: TextField
  private lateinit var blockIntervalField: TextField
  private lateinit var propagationDelayField: TextField
  private lateinit var nodeDegreeField: TextField
  private lateinit var maxBlockSizeField: TextField
  private lateinit var networkBandwidthField: TextField
  private lateinit var transactionADelayField: TextField
  private lateinit var transactionBAccelerationField: TextField
  private lateinit var combinedAttackModeCombo: Combo

  override fun createControl(parent: Composite) {
    val root = Composite(parent, SWT.BORDER)
    GridLayoutFactory.swtDefaults().numColumns(1).applyTo(root)

    val group = Group(root, SWT.NONE)
    group.text = "3SIM Parameters"
    GridLayoutFactory.swtDefaults().numColumns(3).spacing(0, 10).applyTo(group)
    group.layoutData = GridData(SWT.FILL, SWT.BEGINNING, true, false)

    failureThroughputThresholdField = TextField(
      group,
      "Failure Throughput Threshold:",
      " transactions/min",
      DoubleVerifier,
      ThreesimAttributes.FAILURE_THROUGHPUT_THRESHOLD,
      ThreesimAttributes.FAILURE_THROUGHPUT_THRESHOLD_DEFAULT,
      isValueValid = { it.toDoubleOrNull()?.let { it >= MIN_FAILURE_THROUGHPUT_THRESHOLD } ?: false }
    )

    shannonEntropyKField = TextField(
      group,
      "Shannon Entropy K:",
      "",
      DoubleVerifier,
      ThreesimAttributes.SHANNON_ENTROPY_K,
      ThreesimAttributes.SHANNON_ENTROPY_K_DEFAULT,
      isValueValid = { it.toDoubleOrNull()?.let { it > MIN_SHANNON_ENTROPY_K } ?: false }
    )

    nakamotoCoefficientThresholdField = TextField(
      group,
      "Nakamoto Coefficient Threshold:",
      " %",
      DoubleVerifier,
      ThreesimAttributes.NAKAMOTO_COEFFICIENT_THRESHOLD,
      ThreesimAttributes.NAKAMOTO_COEFFICIENT_THRESHOLD_DEFAULT,
      isValueValid = {
        it.toDoubleOrNull()?.let {
          it in MIN_NAKAMOTO_COEFFICIENT_THRESHOLD..MAX_NAKAMOTO_COEFFICIENT_THRESHOLD
        } ?: false
      }
    )

    reliabilityObservationTimespanField = TextField(
      group,
      "Reliability Observation Timespan:",
      " hours",
      DoubleVerifier,
      ThreesimAttributes.RELIABILITY_OBSERVATION_TIMESPAN,
      ThreesimAttributes.RELIABILITY_OBSERVATION_TIMESPAN_DEFAULT,
      isValueValid = { it.toDoubleOrNull()?.let { it > MIN_RELIABILITY_OBSERVATION_TIMESPAN } ?: false }
    )

    blockIntervalField = TextField(
      group,
      "Block Interval:",
      " ms",
      DoubleVerifier,
      ThreesimAttributes.BLOCK_INTERVAL,
      ThreesimAttributes.BLOCK_INTERVAL_DEFAULT,
      isValueValid = { it.toDoubleOrNull()?.let { it > 0 } ?: false }
    )

    propagationDelayField = TextField(
      group,
      "Propagation Delay:",
      " ms",
      DoubleVerifier,
      ThreesimAttributes.PROPAGATION_DELAY,
      ThreesimAttributes.PROPAGATION_DELAY_DEFAULT,
      isValueValid = { it.toDoubleOrNull()?.let { it >= 0 } ?: false }
    )

    nodeDegreeField = TextField(
      group,
      "Node Degree:",
      "",
      LongVerifier,
      ThreesimAttributes.NODE_DEGREE,
      ThreesimAttributes.NODE_DEGREE_DEFAULT,
      isValueValid = { it.toLongOrNull()?.let { it > 0 } ?: false }
    )

    maxBlockSizeField = TextField(
      group,
      "Max Block Size:",
      " bytes",
      LongVerifier,
      ThreesimAttributes.MAX_BLOCK_SIZE,
      ThreesimAttributes.MAX_BLOCK_SIZE_DEFAULT,
      isValueValid = { it.toLongOrNull()?.let { it > 0 } ?: false }
    )

    networkBandwidthField = TextField(
      group,
      "Network Bandwidth:",
      " bytes/s",
      DoubleVerifier,
      ThreesimAttributes.NETWORK_BANDWIDTH,
      ThreesimAttributes.NETWORK_BANDWIDTH_DEFAULT,
      isValueValid = { it.toDoubleOrNull()?.let { it > 0 } ?: false }
    )
    transactionADelayField = TextField(
      group,
      "Transaction A Delay:",
      " ms",
      LongVerifier,
      ThreesimAttributes.TRANSACTION_A_DELAY,
      ThreesimAttributes.TRANSACTION_A_DELAY_DEFAULT,
      isValueValid = { it.toLongOrNull()?.let { it >= 0 } ?: false }
    )

    transactionBAccelerationField = TextField(
      group,
      "Transaction B Acceleration:",
      " ms",
      LongVerifier,
      ThreesimAttributes.TRANSACTION_B_ACCELERATION,
      ThreesimAttributes.TRANSACTION_B_ACCELERATION_DEFAULT,
      isValueValid = { it.toLongOrNull()?.let { it >= 0 } ?: false }
    )

    val combinedAttackLabel = Label(group, SWT.NONE)
    combinedAttackLabel.text = "Combined Attack Mode:"

    combinedAttackModeCombo = Combo(group, SWT.DROP_DOWN or SWT.READ_ONLY)
    combinedAttackModeCombo.setItems(
      "None",
      "Selfish + Race",
      "Selfish + Finney"
    )
    combinedAttackModeCombo.select(0)
    combinedAttackModeCombo.layoutData = GridData(SWT.FILL, SWT.CENTER, true, false)

    val combinedAttackUnitLabel = Label(group, SWT.NONE)
    combinedAttackUnitLabel.text = ""

    combinedAttackModeCombo.addModifyListener {
      updateLaunchConfigurationDialog()
    }

    control = root
    isInitialized = true
  }

  override fun getName(): String {
    return "ATO-SIM Parameters"
  }

  override fun initializeFrom(configuration: ILaunchConfiguration) {
    if (!isInitialized) return

    failureThroughputThresholdField.initializeFrom(configuration)
    shannonEntropyKField.initializeFrom(configuration)
    nakamotoCoefficientThresholdField.initializeFrom(configuration)
    reliabilityObservationTimespanField.initializeFrom(configuration)
    blockIntervalField.initializeFrom(configuration)
    propagationDelayField.initializeFrom(configuration)
    nodeDegreeField.initializeFrom(configuration)
    maxBlockSizeField.initializeFrom(configuration)
    networkBandwidthField.initializeFrom(configuration)
    transactionADelayField.initializeFrom(configuration)
    transactionBAccelerationField.initializeFrom(configuration)
    when (
      configuration.getAttribute(
        ThreesimAttributes.COMBINED_ATTACK_MODE,
        ThreesimAttributes.COMBINED_ATTACK_MODE_DEFAULT
      )
    ) {
      ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_RACE -> combinedAttackModeCombo.select(1)
      ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_FINNEY -> combinedAttackModeCombo.select(2)
      else -> combinedAttackModeCombo.select(0)
    }
  }

  override fun performApply(configuration: ILaunchConfigurationWorkingCopy) {
    if (!isInitialized) return

    failureThroughputThresholdField.performApply(configuration)
    shannonEntropyKField.performApply(configuration)
    nakamotoCoefficientThresholdField.performApply(configuration)
    reliabilityObservationTimespanField.performApply(configuration)
    blockIntervalField.performApply(configuration)
    propagationDelayField.performApply(configuration)
    nodeDegreeField.performApply(configuration)
    maxBlockSizeField.performApply(configuration)
    networkBandwidthField.performApply(configuration)
    transactionADelayField.performApply(configuration)
    transactionBAccelerationField.performApply(configuration)

    val combinedMode = when (combinedAttackModeCombo.selectionIndex) {
      1 -> ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_RACE
      2 -> ThreesimAttributes.COMBINED_ATTACK_MODE_SELFISH_FINNEY
      else -> ThreesimAttributes.COMBINED_ATTACK_MODE_NONE
    }

    configuration.setAttribute(ThreesimAttributes.COMBINED_ATTACK_MODE, combinedMode)
  }

  override fun setDefaults(configuration: ILaunchConfigurationWorkingCopy) {
    if (!isInitialized) return

    failureThroughputThresholdField.setDefaults(configuration)
    shannonEntropyKField.setDefaults(configuration)
    nakamotoCoefficientThresholdField.setDefaults(configuration)
    reliabilityObservationTimespanField.setDefaults(configuration)
    blockIntervalField.setDefaults(configuration)
    propagationDelayField.setDefaults(configuration)
    nodeDegreeField.setDefaults(configuration)
    maxBlockSizeField.setDefaults(configuration)
    networkBandwidthField.setDefaults(configuration)
    transactionADelayField.setDefaults(configuration)
    transactionBAccelerationField.setDefaults(configuration)
    configuration.setAttribute(
      ThreesimAttributes.COMBINED_ATTACK_MODE,
      ThreesimAttributes.COMBINED_ATTACK_MODE_DEFAULT
    )
  }

  override fun activated(workingCopy: ILaunchConfigurationWorkingCopy) {
    super.activated(workingCopy)
    updateLaunchConfigurationDialog()
  }

  override fun isValid(launchConfig: ILaunchConfiguration): Boolean {
    if (!isInitialized) return false

    return failureThroughputThresholdField.isValid() &&
      shannonEntropyKField.isValid() &&
      nakamotoCoefficientThresholdField.isValid() &&
      reliabilityObservationTimespanField.isValid() &&
            blockIntervalField.isValid() &&
            propagationDelayField.isValid() &&
            nodeDegreeField.isValid() &&
            maxBlockSizeField.isValid() &&
            networkBandwidthField.isValid() && transactionADelayField.isValid() &&
            transactionBAccelerationField.isValid()
  }
}